card/mod/01_core/set/all/fetch.rb

```diff
@@ -20,59 +20,70 @@ module ClassMethods
   #     :skip_virtual               Real cards only
   #     :skip_modules               Don't load Set modules
   #     :look_in_trash              Return trashed card objects
+  #     :local_only                 Use only local cache for lookup and storing
   #     new: {  card opts }      Return a new card when not found
   #
-  def fetch mark, opts={}
-    if String === mark
-      case mark
-      when /^\~(\d+)$/ # get by id
-        mark = $1.to_i
-      when /^\:(\w+)$/ # get by codename
-        mark = $1.to_sym
-      end
-    end
-    mark = Card::Codename[mark] if Symbol === mark # id from codename
+  def fetch mark, opts = {}
+    validate_fetch_opts! opts
+    mark = normalize_mark mark
 
     if mark.present?
-      card, mark, needs_caching = fetch_from_cache_or_db mark, opts # have existing
+      card, mark, needs_caching = fetch_from_cache_or_db mark, opts
     else
       return unless opts[:new]
     end
 
-    if Integer===mark
-      return if card.nil? || mark.nil?
-    else
-      return card.renew(opts) if card and card.eager_renew?(opts)
-      if !card or card.type_id==-1 && clean_cache_opts?(opts)       # new (or improved) card for cache
-        needs_caching = true
-        card = new_for_cache mark, opts
-      end
+    if mark.is_a?(Integer)
+      return if card.nil?
+    elsif card && card.new_card? && opts[:new].present?
+      return card.renew(opts)
+    elsif !card || (card.type_unknown? && !skip_type_lookup?(opts))
+      needs_caching = true
+      card = new_for_cache mark, opts # new (or improved) card for cache
     end
 
-    write_to_cache card if Card.cache && needs_caching
+    write_to_cache card, opts if needs_caching
 
     if card.new_card?
-      if opts[:new]
-        return card.renew(opts) if !clean_cache_opts? opts
-      elsif opts[:skip_virtual]
-        return
-      else
-        card.include_set_modules unless opts[:skip_modules]  # need to load modules here to call the right virtual? method
-        return unless card.virtual?
+      case
+      when opts[:new].present? then return card.renew(opts)
+      when opts[:new] # noop for empty hash
+      when opts[:skip_virtual] then return nil
       end
-      card.name = mark.to_s if mark && mark.to_s != card.name
+      card.rename_from_mark mark unless opts[:local_only]
     end
-
+    # need to load modules here to call the right virtual? method
     card.include_set_modules unless opts[:skip_modules]
-    card
+    card if opts[:new] || card.known?
   end
 
-  def fetch_id mark #should optimize this.  what if mark is int?  or codename?
-    card = fetch mark, skip_virtual: true, skip_modules: true
-    card and card.id
+
+
+  def fetch_local mark, opts = {}
+    fetch mark, opts.merge(:local_only=>true)
+  end
+
+  def fetch_id mark
+    if mark.is_a?(Integer)
+      mark
+    elsif mark.is_a?(Symbol) && Card::Codename[mark]
+      Card::Codename[mark]
```

Comment:
shouldn't a symbol always be a codename?  If Card::Codename[mark] is nil, we should return nil, no?